File: wizard/sethlans_wizard/handlers/admin_user.py
# SPDX-FileCopyrightText: 2025 Dryad and Naiad Software LLC
#
# SPDX-License-Identifier: GPL-2.0-or-later
"""``POST /api/wizard/admin-user/`` — Admin user step (FR-M2-5).

Validates the candidate admin tuple via the stdlib password validators
in ``wizard.sethlans_wizard.password_validators`` and stashes the
validated tuple into the in-memory wizard state. The plaintext
password lives in process memory only — it is NEVER written to
``manager.ini``, the wizard log, the launcher log, or any sentinel
until FR-M2-9's pending-setup serialization (FR-PEND4 / NF-6).

Failure responses return HTTP 400 with the union of failure codes —
the caller surfaces them via the page UI. Resource-integrity failure
(common-passwords file missing or hash mismatch) is HTTP 500
``common_passwords_resource_invalid`` per FR-M2-5 fail-closed rule.
"""

from __future__ import annotations

import logging
from pathlib import Path
from typing import Callable, Iterable

from wizard.sethlans_wizard import progress, wizard_state
from wizard.sethlans_wizard.checkpoints import ADMIN_VALIDATED
from wizard.sethlans_wizard.handlers import _wsgi
from wizard.sethlans_wizard.handlers.auth import session_header_valid
from wizard.sethlans_wizard.password_validators import (
    validate_password,
    verify_resource,
)

logger = logging.getLogger(__name__)


def _basic_field_check(payload: dict) -> tuple[str, str, str, str] | str:
    """Pull username/email/password/confirm from *payload*.

    Returns a 4-tuple on success or an error code string on shape
    failure.
    """
    username = payload.get("username")
    email = payload.get("email")
    password = payload.get("password")
    confirm = payload.get("password_confirm")
    for name, value in (
        ("username", username),
        ("email", email),
        ("password", password),
        ("password_confirm", confirm),
    ):
        if not isinstance(value, str) or not value:
            return f"{name}_required"
    if password != confirm:
        return "password_mismatch"
    return username, email, password, confirm  # type: ignore[return-value]


def make_admin_user_handler(data_dir: Path) -> Callable:
    """Return a WSGI handler bound to *data_dir* for FR-M2-5."""
    if not isinstance(data_dir, Path):
        data_dir = Path(data_dir)

    def handler(environ: dict, start_response: Callable) -> Iterable[bytes]:
        return _handle(environ, start_response, data_dir)

    return handler


def _handle(
    environ: dict,
    start_response: Callable,
    data_dir: Path,
) -> Iterable[bytes]:
    method = environ.get("REQUEST_METHOD", "GET").upper()
    if method != "POST":
        return _wsgi.send_json(
            start_response,
            {"error": "Method Not Allowed"},
            status=405,
            extra_headers=[("Allow", "POST")],
        )
    if _wsgi.query_string_has_forbidden_key(environ):
        return _wsgi.send_json(
            start_response,
            {"error": "session token must not appear in URL"},
            status=400,
        )
    if not session_header_valid(environ):
        return _wsgi.send_json(
            start_response,
            {"error": "missing or invalid X-Wizard-Session header"},
            status=401,
        )
    body = _wsgi.read_body(environ)
    if len(body) > _wsgi.BODY_MAX:
        return _wsgi.send_json(
            start_response,
            {"error": "request body too large"},
            status=400,
        )
    payload = _wsgi.parse_json_body(body)
    if payload is None:
        return _wsgi.send_json(
            start_response,
            {"error": "request body must be JSON"},
            status=400,
        )

    # Resource fail-closed (FR-M2-5).
    resource_error = verify_resource()
    if resource_error is not None:
        logger.critical("admin-user handler refused: %s", resource_error)
        return _wsgi.send_json(
            start_response,
            {"error": resource_error},
            status=500,
        )

    field_check = _basic_field_check(payload)
    if isinstance(field_check, str):
        return _wsgi.send_json(
            start_response,
            {"error": field_check},
            status=400,
        )
    username, email, password, _confirm = field_check

    failures = validate_password(password, user_attrs=[username, email])
    if failures:
        # Resource-integrity failures bubble through the validator as
        # ``common_passwords_resource_invalid`` — promote to 500.
        if "common_passwords_resource_invalid" in failures:
            return _wsgi.send_json(
                start_response,
                {"error": "common_passwords_resource_invalid"},
                status=500,
            )
        return _wsgi.send_json(
            start_response,
            {"error": "password_invalid", "failures": failures},
            status=400,
        )

    # Stash the validated tuple. NEVER log the password.
    wizard_state.set_admin(username, email, password)
    progress.append_checkpoint(data_dir, ADMIN_VALIDATED)
    return _wsgi.send_json(
        start_response,
        {"status": "ok", "username": username},
        status=200,
    )


__all__ = ["make_admin_user_handler"]

File: wizard/frontend/static/js/admin_user.js
// SPDX-FileCopyrightText: 2025 Dryad and Naiad Software LLC
//
// SPDX-License-Identifier: GPL-2.0-or-later

// Admin user page (admin-user.html) controller (FR-M2-5).
//
// Inline validation: password mismatch checked client-side; the
// backend's validator rejects (length, similarity, common, all-numeric)
// with HTTP 400 + a `failures` array which we surface verbatim so the
// user sees the exact reasons.
//
// On success, the admin password is POSTed to /api/wizard/admin-user/
// where the wizard backend stashes it in its in-memory wizard_state
// module (FR-M2-5). The worker-password page does NOT receive the
// password back through the browser — when the operator checks "Use
// admin password" there, the worker-password handler reads the admin
// plaintext from wizard_state server-side. This avoids the FE-1
// sessionStorage leak (admin password was previously written to
// `wizard:adminPasswordTransient` which lingered until tab close).
//
// The admin password therefore lives only in this Petite-vue scope's
// `form.password` for the lifetime of the admin-user page; on
// navigation the JS context is torn down and the reference is GC'd.
// NEVER stashed in the regular form-state stash (which is for
// re-display on Back; passwords are excluded per FR-CHK3-FORM-STATE).

import { createApp, reactive } from '/static/vendor/petite-vue.js';
import {
  consumeResumeBanner,
  expireAndRedirect,
  wizardFetch,
} from '/static/js/common.js';
import {
  loadFormState,
  stashFormState,
} from '/static/js/form_state.js';
import {
  STEP_ADMIN,
  applyChromeContext,
  backRouteFor,
  buildStepperModel,
} from '/static/js/wizard_chrome.js';
import {
  evaluatePassword,
  strengthClass,
} from '/static/js/password_strength.js';

const STEP = 'admin-user';

const FAILURE_LABELS = {
  password_too_short: 'Password must be at least 8 characters.',
  password_too_long: 'Password is too long.',
  password_too_similar: 'Password is too similar to the username or email.',
  password_too_common: 'Password is too common.',
  password_all_numeric: 'Password cannot be all digits.',
};

const scope = reactive({
  form: { username: '', email: '', password: '', passwordConfirm: '' },
  submitting: false,
  error: '',
  fieldError: '',
  failures: [],
  resumeBanner: '',
  topology: null,
  // Single eye-toggle state shared by both password fields — clicking
  // either eye reveals/hides BOTH inputs so the user can verify the
  // confirm matches without an extra click.
  showPassword: false,
  stepper: buildStepperModel(
    null, STEP_ADMIN,
    ['topology_chosen', 'network_configured', 'database_configured'],
  ),

  get strength() {
    return evaluatePassword(
      this.form.password, [this.form.username, this.form.email],
    );
  },
  get strengthClassName() {
    const s = this.strength;
    return strengthClass(s.score, s.max);
  },
  // Live confirm-match indicator. Returns null when confirm is empty
  // (so we render nothing) and a boolean once the user has typed.
  get passwordsMatch() {
    if (!this.form.passwordConfirm) return null;
    return this.form.password === this.form.passwordConfirm;
  },

  goBack() {
    const route = backRouteFor(STEP_ADMIN, this.topology);
    if (route) {
      this._stash();
      window.location.assign(route);
    }
  },

  _stash() {
    // Passwords NEVER stashed in the form-state stash.
    stashFormState(STEP, {
      username: this.form.username,
      email: this.form.email,
    });
  },

  async submit() {
    if (this.submitting) return;
    if (!this.form.username || !this.form.email
        || !this.form.password || !this.form.passwordConfirm) {
      this.error = 'Please fill in every field.';
      return;
    }
    if (this.form.password !== this.form.passwordConfirm) {
      this.error = 'Passwords do not match.';
      this.fieldError = 'password_mismatch';
      return;
    }
    this.error = '';
    this.fieldError = '';
    this.failures = [];
    this.submitting = true;
    this._stash();
    let response;
    try {
      response = await wizardFetch('/api/wizard/admin-user/', {
        method: 'POST',
        body: JSON.stringify({
          username: this.form.username,
          email: this.form.email,
          password: this.form.password,
          password_confirm: this.form.passwordConfirm,
        }),
      });
    } catch (_) {
      this.submitting = false;
      this.error = 'Network error. Check the launcher logs and try again.';
      return;
    }
    if (response.status === 401 || response.status === 403) {
      expireAndRedirect(
        'Your session expired — please paste the setup token again.',
      );
      return;
    }
    let payload = null;
    try { payload = await response.json(); } catch (_) { /* tolerate */ }
    if (response.status === 200) {
      // The wizard backend has stashed the admin tuple in its
      // in-memory wizard_state. The browser does NOT round-trip the
      // password to /worker-password — the worker-password handler
      // reads the admin plaintext from wizard_state server-side when
      // the "Use admin password" checkbox is on (FR-M2-6). The
      // worker-password page is always navigated to next; for manager
      // topology that page detects "no worker on this machine" and
      // forwards to /ffmpeg automatically.
      window.location.assign('/worker-password');
      return;
    }
    if (response.status === 500 && payload && payload.error === 'common_passwords_resource_invalid') {
      this.submitting = false;
      this.error = 'The wizard could not load its password validator data. Restart the launcher.';
      return;
    }
    this.submitting = false;
    if (payload && payload.error === 'password_invalid' && Array.isArray(payload.failures)) {
      this.fieldError = 'password_invalid';
      this.failures = payload.failures.map(f => FAILURE_LABELS[f] || f);
      this.error = 'Password did not meet the requirements.';
      return;
    }
    if (payload && typeof payload.error === 'string') {
      this.fieldError = payload.error;
      this.error = 'Please correct the highlighted field.';
      return;
    }
    this.error = 'Could not validate the admin user. Check the launcher logs.';
  },
});

document.addEventListener('DOMContentLoaded', () => {
  const banner = consumeResumeBanner();
  if (banner) scope.resumeBanner = banner;
  const stash = loadFormState(STEP);
  if (stash) {
    if (typeof stash.username === 'string') scope.form.username = stash.username;
    if (typeof stash.email === 'string') scope.form.email = stash.email;
  }
  createApp(scope).mount('#app');
  const f = document.getElementById('admin-username');
  if (f) f.focus();
  applyChromeContext(scope, STEP_ADMIN, [
    'topology_chosen', 'network_configured', 'database_configured',
  ]);
});
